### PR TITLE
Form support

### PR DESCRIPTION
**Pull request by harzzn**

### CODE REVIEW

File: html2wp/setup.php [hunk 1]
```diff
@@ -131,9 +136,19 @@ function html2wp_register_required_plugins() {
 			'source'           => 'https://github.com/html2wp/simple-live-editor/archive/master.zip',
 			'required'         => true,
 			'force_activation' => true,
-		),
+		)
 
 	);
 
+    if ( isset($html2wp_settings["forms"]) && !empty($html2wp_settings["forms"]) ) {
```
Comment (harzzn) [inline html2wp/setup.php]: Please revise https://make.wordpress.org/core/handbook/best-practices/coding-standards/php/#space-usage and https://make.wordpress.org/core/handbook/best-practices/coding-standards/php/#single-and-double-quotes

File: html2wp/setup.php [hunk 2]
```diff
@@ -131,9 +136,19 @@ function html2wp_register_required_plugins() {
...
+    	$plugins[] = array(
+			'name'             => 'Gravity Forms',
+			'slug'             => 'gravityforms',
+			'source'           => 'https://github.com/wp-premium/gravityforms/archive/master.zip',
+			'required'         => true,
+			'force_activation' => true,
+		);
```
Comment (harzzn) [inline html2wp/setup.php:156]: Fix indentation https://make.wordpress.org/core/handbook/best-practices/coding-standards/php/#indentation

File: html2wp/forms.php [hunk 3]
```diff
@@ -0,0 +1,408 @@
+<?php
+/**
+ * Setup the form on activation based on settings
+ *
+ * @package html2wp/simple-wp-starter-theme
+ */
+
+define( 'THEME_DOMAIN', 'html2wp_theme');
+define( 'THEME_DIR', get_template_directory() );
```
Comment (harzzn) [inline html2wp/forms.php]: Remove unused constants

File: html2wp/forms.php [hunk 4]
```diff
@@ -0,0 +1,408 @@
...
+define( 'HTML2WP_FORM_CREATED', 'html2wp_gf_form_created_once_555' );
```
Comment (harzzn) [inline html2wp/forms.php:15]: What is this used for?

File: html2wp/forms.php [hunk 5]
```diff
@@ -0,0 +1,408 @@
...
+define( 'GRAVITY_VERSION', '1.9.17.5' );
+define( 'GRAVITY_PENDING_INSTALLATION', 'gform_pending_installation' );
+define( 'GRAVITY_RG_VERSION_KEY', 'rg_form_version' );
```
Comment (harzzn) [inline html2wp/forms.php:22]: Are these proprietary constants for gravity forms? Please add a comment to code to explain how they are used. Would these be only needed when there is forms?

File: html2wp/forms.php [hunk 6]
```diff
@@ -0,0 +1,408 @@
...
+
+/**
+ * Returns the current version of gravity forms from wp-premium
+ *
+ * @return gf version
+ */
+function html2wp_get_gf_version() {
+    $changelog = trim(file_get_contents("https://raw.githubusercontent.com/wp-premium/gravityforms/master/change_log.txt"));
```
Comment (harzzn) [inline html2wp/forms.php]: Is it safe to read from here if user installs the plugin another way?

File: html2wp/forms.php [hunk 7]
```diff
@@ -0,0 +1,408 @@
...
+ */
+function html2wp_setup_gravity_contact_form() {
+    /**
+     * Get the form data from form config json
+     * in order to create a new GV Form
+     */
+
+    global $html2wp_settings;
+
+    if ( isset($html2wp_settings["forms"]) && !empty($html2wp_settings["forms"]) ) {
```
Comment (harzzn) [inline html2wp/forms.php]: Please revise https://make.wordpress.org/core/handbook/best-practices/coding-standards/php/#space-usage and https://make.wordpress.org/core/handbook/best-practices/coding-standards/php/#single-and-double-quotes

Should this check be a function as it's used in multiple places?

File: html2wp/forms.php [hunk 8]
```diff
@@ -0,0 +1,408 @@
...
+                            'message'           => '{all_fields}',
+                            'from'              => '{admin_email}',
+                            'disableAutoformat' => false,
+                        )
+                    );
+
+                    /**
+                     * Supporting only the default confirmation for now
+                     * 
+                    */
```
Comment (harzzn) [inline html2wp/forms.php]: Remove extra lines

File: html2wp/forms.php [hunk 9]
```diff
@@ -0,0 +1,408 @@
...
+            });         
+
+            //Form has not been created previously, create one now
+            if (empty($form_to_create)) {
+
+                $form = array();
+                $form['title'] = $gf_form_name;
+                $form['gfid'] = $gf_form_id; //custom id for identifying the form
+
+                foreach ($this_form_data["data"] as $key=>$elem) {
```
Comment (harzzn) [inline html2wp/forms.php]: What are we doing here, please add a comment

File: html2wp/forms.php [hunk 10]
```diff
@@ -0,0 +1,408 @@
...
+                }
+            } else {
+            }
+        }
+
+    }
+}
+
+//handle the form submit api endpoint
+if( ! is_admin() ) add_action( 'init', 'html2wp_form_submit_api_endpoint' );
```
Comment (harzzn) [inline html2wp/forms.php]: Add braces https://make.wordpress.org/core/handbook/best-practices/coding-standards/php/#brace-style

File: html2wp/forms.php [hunk 11]
```diff
@@ -0,0 +1,408 @@
...
+        echo "<h3>" . $response[0] . "</h3>";
+        
+        //Footer
+        get_footer();
+
+    }
+}
+
+//Perform setup after this theme is activated
+add_action('after_switch_theme', 'html2wp_setup_theme_components');
```
Comment (harzzn) [inline html2wp/forms.php]: Move html2wp_setup_theme_components and html2wp_detect_plugin_activation to setup.php and merge with existing logic if possible. Also list all actions in the beggining  of the setup.php file.

File: html2wp/forms.php [hunk 12]
```diff
@@ -0,0 +1,408 @@
...
+}
+
+//Perform theme setup after Gravity forms is installed
+add_action( 'activated_plugin', 'html2wp_detect_plugin_activation', 10, 2 );
+
+/**
+ * Peforms contact form setup after Gravity forms plugin is activated
+ *
+ */
+function html2wp_detect_plugin_activation(  $plugin, $network_activation ) {
```
Comment (harzzn) [inline html2wp/forms.php]: Remove unused variable $network_activation

File: html2wp/forms.php [hunk 13]
```diff
@@ -0,0 +1,408 @@
...
+         * as it could a totally different theme.
+         */
+        html2wp_setup_gravity_contact_form();
+        delete_option(GRAVITY_PENDING_INSTALLATION);
+    }
+
+}
+
+//Perform theme setup after Gravity forms is installed
+add_action( 'activated_plugin', 'html2wp_detect_plugin_activation', 10, 2 );
```
Comment (harzzn) [inline html2wp/forms.php]: Remove ", 10, 2" as second variable not needed

File: html2wp/forms.php [hunk 14]
```diff
@@ -0,0 +1,408 @@
...
+         * tables are not created.
+         */
+        GFForms::setup_database();
+
+        /**
+         * Disable the gravity forms installation wizard
+         * as it conflicts with auto setupof forms
+         */
+        update_option(GRAVITY_PENDING_INSTALLATION, -1);     
+        //update_option(GRAVITY_RG_VERSION_KEY, GRAVITY_VERSION);
```
Comment (harzzn) [inline html2wp/forms.php]: Remove commented code

File: html2wp/forms.php [hunk 15]
```diff
@@ -0,0 +1,408 @@
...
+ * after the theme is activated.
+ */
+function html2wp_setup_theme_components () {
+
+    /**
+     * Disable the gravity forms installation wizard
+     * as it conflicts with auto setupof forms
+     */
+    update_option(GRAVITY_PENDING_INSTALLATION, -1);     
+    //update_option(GRAVITY_RG_VERSION_KEY, GRAVITY_VERSION);
```
Comment (harzzn) [inline html2wp/forms.php]: Remove commented code

File: html2wp/forms.php [hunk 16]
```diff
@@ -0,0 +1,408 @@
...
+//handle the form submit api endpoint
+if( ! is_admin() ) add_action( 'init', 'html2wp_form_submit_api_endpoint' );
+
+/**
+ * Handle form submission, this function is called
+ * when wordpress detects the api endpoint called
+ */
+function html2wp_form_submit_api_endpoint() {
+
+    if (strpos($_SERVER['REQUEST_URI'], 'action=html2wp_api') !== false) {
```
Comment (harzzn) [inline html2wp/forms.php]: Why not use the native wp ajax api?

File: html2wp/forms.php (hunk 6 above)
Comment (dkd903) [inline html2wp/forms.php]: This is safe

**Pull request merged**